src/details/ArborX_DetailsDendrogram.hpp

```diff
@@ -0,0 +1,170 @@
+{
+  int source;
+  int target;
+};
+
+// Compute permutation that orders edges in increasing order
+template <typename ExecutionSpace, typename MemorySpace>
+Kokkos::View<unsigned int *, MemorySpace>
+computeEdgesPermutation(ExecutionSpace const &exec_space,
+                        Kokkos::View<WeightedEdge *, MemorySpace> &edges)
+{
+  Kokkos::Profiling::pushRegion("ArborX::Dendrogram::sort_edges");
+
+  int const num_edges = edges.size();
+
+  Kokkos::View<float *, MemorySpace> weights(
+      Kokkos::view_alloc(exec_space, Kokkos::WithoutInitializing,
+                         "ArborX::Dendrogram::weights"),
+      num_edges);
+  Kokkos::parallel_for(
+      "ArborX::Dendrogram::copy_weights",
+      Kokkos::RangePolicy<ExecutionSpace>(exec_space, 0, num_edges),
+      KOKKOS_LAMBDA(int const e) { weights(e) = edges(e).weight; });
+
+  auto permute = Details::sortObjects(exec_space, weights);
+
+  Kokkos::Profiling::popRegion();
+
+  return permute;
+}
+
+template <typename Edges, typename Parents>
+void dendrogramUnionFindHost(Edges sorted_edges_host, Parents &parents_host)
+{
+  using ExecutionSpace = Kokkos::DefaultHostExecutionSpace;
+  ExecutionSpace host_space;
+
+  int const num_edges = sorted_edges_host.size();
+  int const num_vertices = num_edges + 1;
+  auto const vertices_offset = num_edges;
+
+  Kokkos::View<int *, Kokkos::HostSpace> labels(
+      Kokkos::view_alloc(Kokkos::WithoutInitializing,
+                         "ArborX::Dendrogram::dendrogram_union_find::labels"),
+      num_vertices);
+  iota(host_space, labels);
+  UnionFind<Kokkos::HostSpace, true> union_find(labels);
+
+  Kokkos::View<int *, Kokkos::HostSpace> set_edges_host(
+      Kokkos::view_alloc(host_space, Kokkos::WithoutInitializing,
+                         "ArborX::Dendrogram::labels"),
+      num_vertices);
+  constexpr int UNDEFINED = -1;
+  Kokkos::deep_copy(host_space, set_edges_host, UNDEFINED);
+
+  Kokkos::parallel_for(
+      "ArborX::Dendrogram::dendrogram_union_find::union_find",
+      Kokkos::RangePolicy<ExecutionSpace>(host_space, 0, 1),
+      KOKKOS_LAMBDA(int) {
+        for (int e = 0; e < num_edges; ++e)
+        {
+          int i = union_find.representative(sorted_edges_host(e).source);
+          int j = union_find.representative(sorted_edges_host(e).target);
+
+          for (int k : {i, j})
+          {
+            auto edge_child = set_edges_host(k);
+            if (edge_child != UNDEFINED)
+              parents_host(edge_child) = e;
+            else
+              parents_host(vertices_offset + k) = e;
+          }
+
+          union_find.merge(i, j);
+
+          set_edges_host(union_find.representative(i)) = e;
+        }
+        parents_host(num_edges - 1) = -1; // root
+      });
+  Kokkos::Profiling::popRegion();
+}
+
+template <typename ExecutionSpace, typename MemorySpace>
+void dendrogramUnionFind(ExecutionSpace const &exec_space,
+                         Kokkos::View<WeightedEdge *, MemorySpace> edges,
```

Comment:
```suggestion
                         Kokkos::View<WeightedEdge const*, MemorySpace> edges,
```
clarify that this won't change?